sherpa/astro/instrument.py

```diff
@@ -1141,3 +1142,120 @@ def get_kernel(self, data, subkernel=True):
+
+    return DataARF(name, energ_lo=elo, energ_hi=ehi, specresp=specresp, exposure=exposure, ethresh=ethresh)
+
+
+def create_delta_rmf(rmflo, rmfhi, offset=1,
+                     e_min=None, e_max=None, ethresh=None,
+                     name='delta-rmf'):
+
+    _validate_create_rmf_input(rmflo, rmfhi, offset)
+
+    # Set up the delta-function response.
+    # TODO: should f_chan start at startchan?
+    #
+    nchans = rmflo.size
+    matrix = numpy.ones(nchans, dtype=numpy.float32)
+    dummy = numpy.ones(nchans, dtype=numpy.int16)
+    f_chan = numpy.arange(1, nchans + 1, dtype=numpy.int16)
+
+    return sherpa.astro.data.DataRMF(name, detchans=nchans,
+                                     energ_lo=rmflo, energ_hi=rmfhi,
+                                     n_grp=dummy, n_chan=dummy,
+                                     f_chan=f_chan, matrix=matrix,
+                                     offset=offset,
+                                     e_min=e_min, e_max=e_max,
+                                     ethresh=ethresh)
+
+
+def create_non_delta_rmf(rmflo, rmfhi, fname, offset=1,
+                         e_min=None, e_max=None, ethresh=None,
+                         name='delta-rmf'):
+
+    if fname is not None and not os.path.isfile(fname):
+        raise ValueError("{} is not a file".format(fname))
+
+    # Set up the delta-function response.
+    # TODO: should f_chan start at startchan?
+    #
+    nchans = rmflo.size
+
+    n_grp, f_chan, n_chan, matrix = calc_grp_chan_matrix(fname)
+
+    return sherpa.astro.data.DataRMF(name, detchans=nchans,
+                                     energ_lo=rmflo, energ_hi=rmfhi,
+                                     n_grp=n_grp, n_chan=n_chan,
+                                     f_chan=f_chan, matrix=matrix,
+                                     offset=offset,
+                                     e_min=e_min, e_max=e_max,
+                                     ethresh=ethresh)
+
+
+def calc_grp_chan_matrix(fname):
+    try:
+        data, filename = \
+            sherpa.astro.io.backend.get_image_data(fname)
+        matrix = data['y']
+        n_grp = []
+        n_chan = []
+        f_chan = []
+        for row in matrix > 0:
+            flag = numpy.hstack([[0], row, [0]])
+            diffs = numpy.diff(flag, n=1)
+            starts, = numpy.where(diffs > 0)
+            ends, = numpy.where(diffs < 0)
+            n_chan.extend(ends - starts)
+            f_chan.extend(starts + 1)
+            n_grp.append(len(starts))
+        n_grp = numpy.asarray(n_grp, dtype=numpy.int16)
+        f_chan = numpy.asarray(f_chan, dtype=numpy.int16)
+        n_chan = numpy.asarray(n_chan, dtype=numpy.int16)
+        matrix = matrix.flatten()
+        matrix = matrix[matrix > 0]
+        return n_grp, f_chan, n_chan, matrix
+    except sherpa.utils.err.IOErr as ioerr:
+        print(ioerr)
+        raise ioerr
+
+
+def _validate_create_rmf_input(rmflo, rmfhi, startchan, fname=None):
```

Comment:
Given that

1. this routine is not used by `create_non_delta_rmf`
2. I believe `DataRMF` has checks in for this (well, I assume that's what `DataRMF._validate` does, but I haven't got time to check)

is this routine really worth it? It can be nice to get an error "close" to the call site, and that can be specialized for the context of the call site, but I don't know if it is worth it here.

However, we could always remove these checks in a later PR.

Reply:
I think I was trying to get rid of that function altogether and make those checks in the response classes themselves. Let me see if I can still do that or if, as you say, the checks are already there. Of course tests should be doing that, but you know...